Multi stage explain

gortiz opened the pull request:
This PR introduces a new way to explain multi-stage queries in Pinot. 
The main goal is to provide a more detailed explanation of the query execution plan, including information about
the physical operators that are being used.

By default, `explain plan for` will return the new plan. If you want to use the old plan you can use 
`explain plan without implementation for`. This may be problematic, so we can discuss to introduce a new flag for this.
The main reason to break the default behavior is that the new plan is more verbose actually what a user should expect
when asking for _implementation_, at least following Calcite terminology.
Alternatively we can change the syntax in the same way we already did with `explain physical plan for`.

At architectural level, the new explain mode is closer to the one used in single stage.
The broker parses and optimizes the query generating a logical plan, generating RelNodes.
These nodes are transformed into PlanNodes as usual and sent to the servers.
But instead of asking to execute the plan, the broker asks to explain it using a new protobuf endpoint.
This new endpoint returns a list of PlanNodes.

When the server receives the explain request, it analyzes the plan looking for leaf operators and creates single-stage
operators as usual.
There are two key differences with respect of the execution mode:
1. The server tracks which PlanNodes have been converted into single-stage operators.
2. The server does not execute the operator. Instead it calls a new introduced method `Operator.getOperatorInfo`, 
which returns the same information returned by `Operator.explainPlan` but in POJOs.

The server then convert these POJOs into PlanNodes and substitute the tracked PlanNodes with the new ones.
Finally the new plan is sent back to the broker.

In order to be able to introduce physical (aka index used, etc) information in the PlanNode, a new ExplainedPlanNode is
created.
These nodes are not meant to be translated into actual operators, but to be used to explain the query execution plan.
When the broker receives the PlanNodes, it converts them back into a RelNode using a new class `PinotExplainedRelNode`.

Then the broker substitutes the original logical RelNodes with the new ones returned by the servers.
Finally, it explains the RelNode as expected in Calcite.

The result can be see in the following pictures:

Without implementation (similar to current explain)
![image](https://github.com/user-attachments/assets/761ee0d5-24d5-4f10-8d1c-d9c4a886ebef)


With implementation:
![image](https://github.com/user-attachments/assets/13e96262-3ea4-4479-89a9-7c201dfe8993)


The PR is still a work in progress, but it is already partially functional.

- [x] Combine different plans for segments (right now the first one is used)
    - Ideally we should _group by plan_ and count how many segments use that plan
- [x] Change behavior when verbose explain mode is used
   - Right now verbose mode is ignored.
- [x] Add a flag so that we can use to decide if we want to use the new plan or the old one by default
- [x] PlanNodeToRelConverted
  - [x] Support Window
  - [x] Support SetOp
  - [x] Decide what to do with send and receive mailboxes
- [x] Support pipeline breaker
  - Given current explain does not support pipeline breaker, we decided to do not support pipeline breaker in the first version of the new explain
- [x] PinotTable.getRowType may not use TypeFactory
- [x] Try to simplify LeafStageTransferableBlockOperator.explain
- [x] Try to simplify QueryServer.explain (similar to what was done in QueryDispatcher)
- [x] Try to simplify QueryRunner.explain (similar to what was done in QueryDispatcher)
- [x] Resolve corner cases in MultiStageBrokerRequestHandler
- [x] Review ServerQueryExecutorV1Impl
- [x] Decide if we need to keep MultiStageExplainAskingServersUtils
- [x] Verify tests
- [x] Split PinotExplainedRelNode and PinotExplainedRelNode.Info
  - Edit: PinotExplainedRelNode.Info was renamed as ExplainInfo and moved out from PinotExplainedRelNode

gortiz commented on pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java:
```diff
@@ -1192,6 +1209,9 @@ public enum WindowOverFlowMode {
     }
+
+    public static final String ASK_SERVERS_FOR_EXPLAIN_PLAN = "pinot.query.explain.ask.servers";
+    public static final boolean DEFAULT_ASK_SERVERS_FOR_EXPLAIN_PLAN = true;
```
Probably this should be changed to false in the initial version. I'm keeping it as true for a while to verify that all tests are running.

yashmayya reviewed (commented):
I'm only through about half this PR but this is a really cool feature that is gonna be super useful! I'll review the remaining bits in a subsequent pass soon.

gortiz commented on pinot-common/src/main/proto/plan.proto:
```diff
@@ -193,3 +194,29 @@ message WindowNode {
 }
+
+message ExplainNode {
+  string type = 1;
```
Yep, this is basically the _type_ of operator, which may be anything. Remember that these operators can even be extended with plugable indexes. Anyway we can add a comment to make it clearer for future readers :+1:

gortiz replied:
I think I'm going to rename it as _title_. WDYT?

gortiz replied:
I think title is a better name. Type seems to indicate that it is a unique thing that completely identifies the operator, when in fact it may not be. I've added some documentation which emphasizes that explain nodes should not be used to reconstruct the operators.

gortiz commented on pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java:
```diff
@@ -1192,6 +1209,9 @@ public enum WindowOverFlowMode {
       public static final int V1 = 1;
     }
+
+    public static final String ASK_SERVERS_FOR_EXPLAIN_PLAN = "pinot.query.explain.ask.servers";
```
I prefer to do not use _physical_ given we use that term to refer to different things (in Calcite, in distribution, here, etc). But I think it makes sense to make more clear that it is only used in multi-stage

gortiz replied:
Changed to `pinot.query.multistage.explain.include.segment.plan`. I prefer to try to use the property name as a classifier. First `pinot`, then `query`, then `multistage` and `explain`

gortiz commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/MultiStageOperator.java:
```diff
@@ -167,6 +173,26 @@ protected TransferableBlock updateEosBlock(TransferableBlock upstreamEos, StatMa
+    return new ExplainInfo(getExplainName(), getExplainAttributes(), getChildrenExplainInfo());
+  }
+
+  protected List<ExplainInfo> getChildrenExplainInfo() {
+    return getChildOperators().stream()
+        .filter(Objects::nonNull)
+        .map(Operator::getExplainInfo)
+        .collect(Collectors.toList());
+  }
+
+  protected String getExplainName() {
+    return toExplainString();
+  }
+
+  protected Map<String, Plan.ExplainNode.AttributeValue> getExplainAttributes() {
+    return Collections.emptyMap();
+  }
```
That is what I tried on the first iteration but... these method should be public, which is something I don't like because that is not part of the `Operator` interface. We could create another abstract class that extended by both `BaseOperator` and `MultiStageOperator`. I'm open to add that change, but I think this is simpler for now.

gortiz commented on pinot-query-planner/src/main/java/org/apache/pinot/query/planner/explain/PhysicalExplainPlanVisitor.java:
```diff
@@ -248,6 +249,11 @@ public StringBuilder visitValue(ValueNode node, Context context) {
 
+  @Override
+  public StringBuilder visitExplained(ExplainedNode node, Context context) {
+    return appendInfo(node, context);
```
That is the case in this PR. If in the future we support asking servers in this mode, we will need to modify this code.

gortiz commented on pinot-query-planner/src/main/java/org/apache/pinot/query/planner/explain/ExplainNodeSimplifier.java:
```diff
@@ -0,0 +1,169 @@
+        PlanNode merged = PlanNodeMerger.mergePlans(child1, child2, false);
+        if (merged == null) {
+          LOGGER.info("Found unmergeable inputs on node of type {}: {} and {}", node, child1, child2);
+          assert false : "Unmergeable inputs found";
```
We shouldn't reach this code. All nodes at this point should be mergeable. In case we end up reaching this code in production, it would be better to not fail. But during tests we should to make it easier to find the problem.

gortiz commented on pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java:
```diff
@@ -691,4 +662,132 @@ private void addPrunerStats(InstanceResponseBlock instanceResponse, SegmentPrune
+      ExplainV2ResultBlock block = new ExplainV2ResultBlock(queryContext, explainInfo);
+
+      instanceResponse = new InstanceResponseBlock(block);
+      // TODO: Study if this metadata can/should be added
```
Do not forget about this TODO

gortiz commented on pinot-core/src/main/java/org/apache/pinot/core/plan/PinotExplainedRelNode.java:
```diff
@@ -0,0 +1,146 @@
+    _inputs.set(ordinalInParent, p);
+  }
+
+  @Override
+  public String getRelTypeName() {
+    return _type;
+  }
```
We need to override this method to because the result is name that is presented in the explain plan. If we don't override it, the explain plan of all these nodes would be `PinotExplainedRelNode` for all cases.

gortiz commented on pinot-query-planner/src/main/java/org/apache/pinot/query/catalog/PinotTable.java:
```diff
@@ -44,8 +43,13 @@ public PinotTable(Schema schema) {
   public RelDataType getRowType(RelDataTypeFactory relDataTypeFactory) {
-    Preconditions.checkState(relDataTypeFactory instanceof TypeFactory);
-    TypeFactory typeFactory = (TypeFactory) relDataTypeFactory;
+    // TODO: Look for a better solution
+    TypeFactory typeFactory;
+    if (relDataTypeFactory instanceof TypeFactory) {
+      typeFactory = (TypeFactory) relDataTypeFactory;
+    } else {
+      typeFactory = TypeFactory.INSTANCE;
+    }
```
TBH I don't remember exactly, but it looks like this method can be called with a RelDataTypeFactory that may not be a TypeFactory, in which case it fails to generate the row types.

I was trying to reproduce the issue and I wasn't able to do it, so maybe it was something I've got while I have the partial code? Given this explain plan will be disabled by default, we can remove this change and go back to it if is actually needed.

gortiz commented on pinot-query-planner/src/main/java/org/apache/pinot/query/QueryEnvironment.java:
```diff
@@ -172,7 +185,27 @@ public QueryPlannerResult explainQuery(String sqlQuery, SqlNodeAndOptions sqlNod
             explain.getDetailLevel() == null ? SqlExplainLevel.DIGEST_ATTRIBUTES : explain.getDetailLevel();
         Set<String> tableNames = RelToPlanNodeConverter.getTableNamesFromRelRoot(relRoot.rel);
-        return new QueryPlannerResult(null, PlannerUtils.explainPlan(relRoot.rel, format, level), tableNames);
+        if (!explain.withImplementation() || !askServers) {
```
The `WITH IMPLEMENTATION` is part of Calcite. The idea is that `WITHOUT IMPLEMENTATION` should show the logical (LogicalFilter, etc) plan while by default or with `WITH IMPLEMENTATION` should return the physical plan (the actual operator we use, something closer to what we are showing in this PR).

IMHO using `EXPLAIN IMPLEMENTATION PLAN FOR` was a mistake. We should have created something like `EXPLAIN PLAN WITH <our extension>` where `<our extension` is something like `TOPOLOGY` or `DISTRIBUTION` or something like that. We can try to use this PR as a base for a future extension where we can include distribution information in the normal plan (ie we could include, in mailbox send, the fanout as we do in execution stats)

gortiz commented on pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineExplainIntegrationTest.java:
```diff
@@ -0,0 +1,255 @@
+    explain("SELECT 1 FROM mytable",
+        //@formatter:off
+        "Execution Plan\n"
+            + "PinotLogicalExchange(distribution=[broadcast])\n"
```
I thought about that and preferred to keep it like this just to clearly mark where the synthetic nodes start, but I'm open to change it.

gortiz commented on pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineExplainIntegrationTest.java:
```diff
@@ -0,0 +1,255 @@
+    //@formatter:on
+  }
+
+  @Override
+  protected String getTableName() {
+    return _tableName;
+  }
```
I just copied it from MultiStageEngineIntegrationTest, which was actually changing the value :D. I'm going to remove it.

gortiz commented on pinot-core/src/main/java/org/apache/pinot/core/query/request/context/utils/QueryContextConverterUtils.java:
```diff
@@ -155,11 +157,23 @@ public static QueryContext getQueryContext(PinotQuery pinotQuery) {
+      explainMode = ExplainMode.NONE;
+    } else {
+      String useMultistageEngine = CommonConstants.Broker.Request.QueryOptionKey.USE_MULTISTAGE_ENGINE;
+      if (pinotQuery.getQueryOptions() != null && pinotQuery.getQueryOptions().containsKey(useMultistageEngine)) {
```
Fixed in 34c6eca87595f9564d37ff6ee7dc4144e96a2168

gortiz commented on pinot-core/src/main/java/org/apache/pinot/core/query/request/context/QueryContext.java:
```diff
@@ -528,7 +538,16 @@ public Builder setExpressionOverrideHints(Map<ExpressionContext, ExpressionConte
+    /**
+     * @deprecated Use {@link #setExplain(ExplainMode)} instead.
+     */
+    @Deprecated
```
I just wanted to be more polite in case someone is using it. I don't think it hurts and we can remove it later.

gortiz commented on pinot-core/src/main/java/org/apache/pinot/core/plan/PinotExplainedRelNode.java:
```diff
@@ -0,0 +1,146 @@
+  public RelWriter explainTerms(RelWriter pw) {
+    RelWriter relWriter = super.explainTerms(pw);
+    int inputSize = getInputs().size();
+    if (inputSize == 1) {
+      relWriter.input("input", getInputs().get(0));
+    } else if (inputSize > 1) {
+      for (int i = 0; i < inputSize; i++) {
+        relWriter.input("input#" + i, getInputs().get(i));
+      }
+    }
```
> Where is this input term string being used?

It is used internally in RelWriter (in Calcite).

> Does it matter if for a single input node we use input#0 instead of input?

Not really. As far as I can tell the only requirement is that inputs should have a unique name. So each input should be have a different `input#<id>`. How this name is used depends on the Calcite printer and as far as I can see neither of them actually print this name

gortiz commented on pinot-core/src/main/java/org/apache/pinot/core/plan/PinotExplainedRelNode.java:
```diff
@@ -0,0 +1,146 @@
+  private final List<RelNode> _inputs;
+  private final DataSchema _dataSchema;
+
+  public PinotExplainedRelNode(RelOptCluster cluster, RelTraitSet traitSet, String type,
```
Removed

gortiz commented on pinot-core/src/main/java/org/apache/pinot/core/plan/PinotExplainedRelNode.java:
```diff
@@ -0,0 +1,146 @@
+  private final List<RelNode> _inputs;
+  private final DataSchema _dataSchema;
+
+  public PinotExplainedRelNode(RelOptCluster cluster, RelTraitSet traitSet, String type,
```
Added

gortiz commented on pinot-common/src/main/proto/plan.proto:
```diff
@@ -193,3 +194,29 @@ message WindowNode {
+      string string = 2;
+      int64 long = 3;
+      bool bool = 4;
+      string json = 5;
```
Changed with StringList

gortiz commented on pinot-common/src/main/proto/plan.proto:
```diff
@@ -193,3 +194,29 @@ message WindowNode {
+      DEFAULT = 0;
+      // When being merge, values must be the same. Otherwise cannot be merged.
+      IDEMPOTENT = 1;
+      // When being merge, if values are different, the value can be ignored.
```
Done

gortiz commented on pinot-core/src/main/java/org/apache/pinot/core/operator/query/FilteredGroupByOperator.java:
```diff
@@ -230,4 +232,26 @@ public String toExplainString() {
+          .map(AggregationFunction::toExplainString)
+          .collect(Collectors.toList());
+      attributeBuilder.putJson("aggregations", aggregations);
+    }
```
Added  `_aggregationFunctions`

gortiz commented on pinot-core/src/main/java/org/apache/pinot/core/operator/query/AggregationOperator.java:
```diff
@@ -100,4 +103,21 @@ public String toExplainString() {
+    if (_aggregationFunctions.length == 0) {
+      return;
+    }
+    List<String> aggregations = Arrays.stream(_aggregationFunctions)
+        .map(AggregationFunction::toExplainString)
+        .collect(Collectors.toList());
+    attributeBuilder.putLongIdempotent("numAggregations", aggregations.size());
```
Solved. Now it includes the whole list

gortiz commented on pinot-core/src/main/java/org/apache/pinot/core/operator/ProjectionOperator.java:
```diff
@@ -100,6 +100,17 @@ public String toExplainString() {
+
+  @Override
+  protected void explainAttributes(ExplainAttributeBuilder attributeBuilder) {
+    super.explainAttributes(attributeBuilder);
```
Ideally all explainAttributes should call its super just in case in future we add something common in the super class.

gortiz replied:
In fact I've just added that in two operators that weren't doing it

gortiz commented on pinot-common/src/main/java/org/apache/pinot/common/utils/DatabaseUtils.java:
```diff
@@ -32,6 +32,10 @@ public class DatabaseUtils {
   private DatabaseUtils() {
   }
 
+  public static String[] splitTableName(String tableName) {
```
Added the javadoc. The reason to add that is because Calcite RelBuilder supports qualified table names. In case we use them, this should honor them.

Anyway I don't think PlanNodeToRelConverted.visitTableScan will ever be called given TableScans should always be transformed into other operators in the leaf stage.

gortiz commented on pinot-common/src/main/java/org/apache/pinot/common/utils/DataSchema.java:
```diff
@@ -196,27 +201,135 @@ public int hashCode() {
+    FLOAT(NullValuePlaceHolder.FLOAT) {
+      @Override
+      public RelDataType toType(RelDataTypeFactory typeFactory) {
+        return typeFactory.createSqlType(SqlTypeName.FLOAT);
```
Changed. I'm not sure if we can unify these methods. It would be great, but I think we can do that in a future PR.

yashmayya commented on pinot-query-planner/src/main/java/org/apache/pinot/query/planner/explain/PlanNodeSorter.java:
```diff
@@ -0,0 +1,228 @@
+ * Any comparator can be passed to the sort method to sort the plan nodes, although the default comparator
+ * is used to sort the plan nodes based on the type and the attributes of the node.
+ */
+public class PlanNodeSorter {
```
I didn't quite get the purpose of this plan node sorter utility. In which cases would we get a non-deterministic order of plan nodes (necessitating this plan node sorter)?

gortiz replied:
We receive different plans from different servers and each server generate a different plan for each segment. The order inside each server is not guaranteed (they are calculated concurrently). Once we receive the server plans in the broker, we first sort each server plan.

Once we have each server plan sorted, we merge them easily. Not only that, the result will always be shown in the same, which will help users. Otherwise two explains of the same query on the same data will return plans that are semantically equal but the order of some nodes will be different.

Remember this sorting is only applied to `Combine` explain nodes. In other cases reordering inputs may break semantics.

Two notes:
1. This is something we also do in single-stage explain (see ExplainPlanDataTableReducer). In fact I copied the definition from there. As you can imagine, just sorting the nodes whose name contains `Combine` is not formally correct, but seems to be good enough to be used for a while.
2. Your comment make me realize the code right now is looking for `COMBINE` in the text instead of the new upper camel case `Combine`. I'm changing that in a new commit.

yashmayya commented on pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/RexExpressionUtils.java:
```diff
@@ -28,27 +28,174 @@
+    for (RexExpression functionOperand : functionOperands) {
+      operands.add(toRexNode(builder, functionOperand));
+    }
+    // TODO: This needs to be improved.
```
What exactly needs to be improved / changed here? Also in `getAggFunction`.

yashmayya commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/QueryRunner.java:
```diff
@@ -256,4 +262,66 @@ private Map<String, String> consolidateMetadata(Map<String, String> customProper
+    StageMetadata stageMetadata = stagePlan.getStageMetadata();
+    Map<String, String> opChainMetadata = consolidateMetadata(stageMetadata.getCustomProperties(), requestMetadata);
+
+    if (PipelineBreakerExecutor.hasPipelineBreakers(stagePlan)) {
+      // TODO: Support pipeline breakers before merging this feature.
+      LOGGER.error("Pipeline breaker is not supported in explain query");
+      return stagePlan;
+    }
```
I guess we're not going to support queries with pipeline breakers with this new explain plan initially right? What do we want to show in those cases?

gortiz replied:
I modified the description yesterday to indicate the PipelineBreaker state. Basically we in the current explain we just prune these nodes. In the new explain we will fail with that error shown here.

We can work on that feature in the future.

gortiz replied:
The main problem is that in order to have the exact physical plan in this case we would need to actually execute the pipeline breaker part. For example, a query like:

```sql
select whatever from table1 
where col1 in (select something from table2 where col2 = cte)
```

the actual physical plan will depend on the result of `(select something from table2 where col2 = cte)`. Assuming that subquery is evaluated to [100, 200, 300], the query would be:

```sql
select whatever from table1 
where col1 in (100, 200, 300)
```

Which could, for example, use a inverted index. I had the happy idea of generating a random set of values on the `col1` type, but that could create incorrect plans. For example imagine we randomly generate the set of values `1,2`. The query we would be using to explain would be:

```sql
select whatever from table1 
where col1 in (1, 2)
```

Now imagine that values of `col1` went from 100 to 200. That query would be explained as `ALL_SEGMENTS_PRUNED_ON_SERVER` when the actual plan would be probably different

yashmayya replied:
Ah, I hadn't thought of that either, thanks for the explanation! I guess we can just update that TODO comment for now, it makes sense to defer this considering the current explain also doesn't properly support pipeline breaker.

yashmayya commented on pinot-core/src/main/java/org/apache/pinot/core/operator/query/SelectionPartiallyOrderedByAscOperator.java:
```diff
@@ -77,6 +78,6 @@ public int getNumDocsScanned() {
   @Override
   protected String getExplainName() {
-    return EXPLAIN_NAME;
+    return CaseFormat.UPPER_UNDERSCORE.to(CaseFormat.UPPER_CAMEL, EXPLAIN_NAME);
```
Comment can be resolved, fixed in https://github.com/apache/pinot/pull/13733/commits/9339792aa9856d606afd3cec6513e53e55b4bdb3.

gortiz commented on pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/PlanNodeToRelConverter.java:
```diff
@@ -0,0 +1,468 @@
+                .collect(Collectors.toList());
+
+        _builder.aggregate(groupKey, aggCalls);
+      } catch (RuntimeException e) {
+        LOGGER.warn("Failed to convert aggregate node: {}", node, e);
+        _builder.push(new PinotExplainedRelNode(_builder.getCluster(), "UnknownAggregate", Collections.emptyMap(),
+            node.getDataSchema(), readAlreadyPushedChildren(node)));
+      }
```
Given the builder is quite complex, some subtle conditions may produce this error. Given the class javadoc explicitly that the generated tree should be used in explains because semantics may be lost, I think it is fair to return this synthetic node. In the end what we are trying to give users is something like:

```
Execution Plan
LogicalSort(offset=[0], fetch=[10])
  PinotLogicalSortExchange(distribution=[hash], collation=[[]], isSortOnSender=[false], isSortOnReceiver=[false])
    UnknownAggregate
      LogicalJoin(condition=[=($3, $5)], joinType=[inner])
        ...
```

Instead of an error. Being optimistic, I think this will be more useful than just having a blind error.

gortiz commented on pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/PlanNodeToRelConverter.java:
```diff
@@ -0,0 +1,468 @@
+    return visitor.build();
+  }
+
+  private static class ConverterVisitor implements PlanNodeVisitor<Void, Void> {
```
It is similar to the other visitor I've converted. The fact that the implementation uses a visitor is an implementation detail we don't need to leak into the caller. Imagine the case you mention. Would you be able to know if you can reuse the same visitor in two different calls? Why would you need to call builder.build() in your code? What is easier to read?

```java
Builder builder = ...;
PlanNode planNode = ...;
RelNode relNode = PlanNodeToRelConverter.convert(builder, planNode);
```

or

```java
Builder builder = ...;
PlanNode planNode = ...;
PlanNodeToRelConverter visitor = new PlanNodeToRelConverter.convert(builder, planNode);
planNode.visit(visitor, null);
RelNode relNode = builder.build();
```

Also... the `convert` method has no javadoc but in case we have to add it, it is simpler to add it there.

I love visitors (although I would love even more to be able to use sealed classes) but they are a poor man switch. Switches are great tools to use in your functions, but in Java we offer methods to be called, not switches.

yashmayya replied:
Makes sense, thanks for elaborating! 😄

gortiz commented on pinot-core/src/main/java/org/apache/pinot/core/query/request/context/utils/QueryContextConverterUtils.java:
```diff
@@ -155,11 +157,26 @@ public static QueryContext getQueryContext(PinotQuery pinotQuery) {
+        .setExplain(explainMode).build();
+  }
+
+  private static boolean isUsingV1(PinotQuery pinotQuery) {
```
Renamed and modified to delegate on QueryOptionUtils. I'm not directly using isUseMultistageEngine because I assume the map may be null, so we need to verify that first

gortiz commented:
I've refactored the code a bit to reduce the usage of generic java.util.Function and add instead a AskingServerStageExplainer. I've also added some javadocs. I hope this will make it easier to understand this PR.

gortiz commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/service/dispatch/DispatchClient.java:
```diff
@@ -50,12 +51,17 @@ public ManagedChannel getChannel() {
+
+  public void explain(Worker.QueryRequest request, QueryServerInstance virtualServer, Deadline deadline,
+      Consumer<AsyncResponse<List<Worker.ExplainResponse>>> callback) {
+    _dispatchStub.withDeadline(deadline).explain(request, new AllValuesDispatchObserver<>(virtualServer, callback));
```
I don't know why we only keep the last message in the submit case, but you can see the older implementation is using [DispatchObserver](https://github.com/apache/pinot/blob/master/pinot-query-runtime/src/main/java/org/apache/pinot/query/service/dispatch/DispatchObserver.java) and verify that it is keeping only the latest `QueryResponse` received in `onNext`, which is what `LastValueDispatcherObserver` does.

TBH the old behavior is correct. You can see that `org.apache.pinot.query.service.server.QueryServer.submit` does only call `onNext` once, so it is ok to keep the last value. What I don't get is why the proto file is defined as:
```
  rpc Submit(ServerRequest) returns (stream ServerResponse);
```
instead of
```
  rpc Submit(ServerRequest) returns (ServerResponse);
```

Anyway, if we change the server implementation we will need to use the new `AllValuesDispatcherObserver` and change the code a bit to work with a list instead of a single element.

yashmayya replied:
> What I don't get is why the proto file is defined as:
>
>  ```rpc Submit(ServerRequest) returns (stream ServerResponse);```
> instead of
>
> ```  rpc Submit(ServerRequest) returns (ServerResponse);```

That's the proto definition for [GrpcQueryServer](https://github.com/apache/pinot/blob/de577bc457b580e89ecb4e82076ce09f209bca18/pinot-core/src/main/java/org/apache/pinot/core/transport/grpc/GrpcQueryServer.java#L65) right (which is for v1 streaming queries FWICT)? The proto definition for the multi-stage engine's [QueryServer](https://github.com/apache/pinot/blob/de577bc457b580e89ecb4e82076ce09f209bca18/pinot-query-runtime/src/main/java/org/apache/pinot/query/service/server/QueryServer.java#L50) `Submit` RPC is - https://github.com/apache/pinot/blob/de577bc457b580e89ecb4e82076ce09f209bca18/pinot-common/src/main/proto/worker.proto#L26

So I guess this makes sense now since the new `Explain` RPC returns a `stream ExplainResponse` and the implementation does call `onNext` multiple times.

gortiz replied:
You are right, the Submit endpoint is not defined as stream, so it makes sense to only keep the last (as only 1 is expected).

BTW I don't think submit is used for V1. Remember we have two classes called QueryServer. `org.apache.pinot.query.service.server.QueryServer` is the one used in V2 and `org.apache.pinot.core.transport.QueryServer` is the one used in V1. Then we have ` org.apache.pinot.core.transport.grpc.GrpcQueryServer`. The later is used to execute V1 queries using GPRC and implements the protocol defined in `server.proto`.

gortiz commented on pinot-core/src/main/java/org/apache/pinot/core/operator/transform/TransformOperator.java:
```diff
@@ -86,7 +88,17 @@ public String toExplainString() {
 
   protected String getExplainName() {
-    return EXPLAIN_NAME;
+    return CaseFormat.UPPER_UNDERSCORE.to(CaseFormat.UPPER_CAMEL, EXPLAIN_NAME);
```
Yes. Good call. Luckily a test detected the same problem. The code is changed now and `toExplainString` (the one used in V1) does not call `getExplainName` (the one used in V2). Names are very confusing btw.

yashmayya reviewed (commented):
@gortiz something I just realized is that we can lose the table related information in the new explain plan. For instance, this query on the basic quickstart:

```
EXPLAIN PLAN WITHOUT IMPLEMENTATION FOR WITH tmp AS (
  select playerID,
    teamID,
    SUM(homeRuns) as totalHomeRuns
  from baseballStats
  WHERE yearID > 2000
  GROUP BY playerID,
    teamID
  ORDER BY totalHomeRuns DESC
)
SELECT *
FROM tmp
  JOIN dimBaseballTeams ON tmp.teamID = dimBaseballTeams.teamID;
```

returns:

```
Execution Plan
LogicalJoin(condition=[=($1, $3)], joinType=[inner])
  PinotLogicalExchange(distribution=[hash[1]])
    PinotLogicalAggregate(group=[{0, 1}], agg#0=[$SUM0($2)])
      PinotLogicalExchange(distribution=[hash[0, 1]])
        LeafStageCombineOperator
          StreamingInstanceResponse
            CombineGroupBy
              GroupBy(groupKeys=[[playerID, teamID]], aggregations=[[sum(homeRuns)]])
                Project(columns=[[homeRuns, teamID, playerID]])
                  DocIdSet(maxDocs=[10000])
                    FilterFullScan(predicate=[yearID > '2000'], operator=[RANGE])
  PinotLogicalExchange(distribution=[hash[0]])
    LeafStageCombineOperator
      StreamingInstanceResponse
        StreamingCombineSelect
          SelectStreaming(segment=[dimBaseballTeams_OFFLINE_0], table=[dimBaseballTeams], totalDocs=[51])
            Project(columns=[[teamName, teamID]])
              DocIdSet(maxDocs=[10000])
                FilterMatchEntireSegment(numDocs=[51])
```

whereas earlier it would've returned:

```
Execution Plan
LogicalJoin(condition=[=($1, $3)], joinType=[inner])
  PinotLogicalExchange(distribution=[hash[1]])
    PinotLogicalAggregate(group=[{0, 1}], agg#0=[$SUM0($2)])
      PinotLogicalExchange(distribution=[hash[0, 1]])
        PinotLogicalAggregate(group=[{16, 25}], agg#0=[$SUM0($11)])
          LogicalFilter(condition=[>($27, 2000)])
            LogicalTableScan(table=[[default, baseballStats]])
  PinotLogicalExchange(distribution=[hash[0]])
    LogicalProject(teamID=[$3], teamName=[$4])
      LogicalTableScan(table=[[default, dimBaseballTeams]])
```

For simpler queries, this probably won't be a big issue but for complex queries with lots of joins, CTEs etc. I think it's pretty important to include this information.

Looks like you updated `StreamingSelectionOnlyOperator` to include the table name in its explain attributes; I guess we should do something similar for all other such operators that can be used in the leaf stage (for instance, `StreamingInstanceResponseOperator`)?

gortiz commented:
> something I just realized is that we can lose the table related information in the new explain plan. For instance, this query on the basic quickstart:
> Looks like you updated StreamingSelectionOnlyOperator to include the table name in its explain attributes; I guess we should do something similar for all other such operators that can be used in the leaf stage (for instance, StreamingInstanceResponseOperator)?

Nice catch. Yes, I've found the same problem and tried to solve it that way, but it doesn't seem to be a scalable solution. Very easily we can end up having an operator that is not registering the table. Instead what I've done is to add the table in the `LeafStageTransferableBlockOperator`, which by definition knows the table

yashmayya commented on pinot-query-planner/src/main/java/org/apache/pinot/query/planner/explain/PlanNodeSorter.java:
```diff
@@ -0,0 +1,229 @@
+          return cmp2;
+        }
+        int cmp3;
+        switch (value1.getValueCase()) {
```
Do we not need to handle the `STRINGLIST` case here?

gortiz replied:
Probably because either I missed it or because JSON (the previous type) is difficult to sort. Adding it now that it is just a list of strings

gortiz replied:
Changed in  https://github.com/apache/pinot/pull/13733/commits/272263125a022bf5419ca85c71d766f6f976a83a

yashmayya commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/LeafStageTransferableBlockOperator.java:
```diff
@@ -270,6 +276,69 @@ private TransferableBlock constructMetadataBlock() {
+    } else {
+      attributes = Collections.singletonMap("table", Plan.ExplainNode.AttributeValue.newBuilder()
+          .setString(tableName)
+          .setMergeType(Plan.ExplainNode.AttributeValue.MergeType.IDEMPOTENT)
```
This is the default merge type for string attribute values right?

gortiz replied:
No, the default is _DEFAULT_ :laughing:. But you are right, semantically is the same for Strings and given how GRPC works, it would reduce a couple of bytes per operator

yashmayya reviewed (approved):
Thanks @gortiz, LGTM! There are some new tests added in https://github.com/apache/pinot/pull/13999 failing with an NPE here though (and looks related to the explain changes).

gortiz commented:
The issue should be fixed. Anyway, my error merging this code shown that https://github.com/apache/pinot/pull/13999 is not testing the empty case and also to me it looks like the code that was added in `ServerQueryExecutorV1Impl` in that PR should have been added into `ResultsBlockUtils.buildEmptyQueryResults`. @ankitsultana can you take a look at that?

yashmayya commented on pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java:
```diff
@@ -391,6 +391,14 @@ public static class QueryOptionKey {
+         *
+         * Use false in order to mimic behavior of Pinot 1.2.0 and previous.
+         */
+        public static final String EXPLAIN_ASKING_SERVERS = "explainAskingServers";
```
I think we should update the query option for consistency since we updated the broker config from `pinot.query.explain.ask.servers ` to `pinot.query.multistage.explain.include.segment.plan`. Probably to `explainIncludeSegmentPlan`?

yashmayya commented on pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageBrokerRequestHandler.java:
```diff
@@ -82,6 +84,7 @@ public class MultiStageBrokerRequestHandler extends BaseBrokerRequestHandler {
 
   private final WorkerManager _workerManager;
   private final QueryDispatcher _queryDispatcher;
+  private final boolean _explainAskingServerDefault;
```
nit: let's update this member variable name too.

yashmayya commented:
https://github.com/apache/pinot/pull/14193
